File: JuLayout/OCLayout/UIView+JuLayout.m
//
//  UIView+Layout.m
//  testBlock
//
//  Created by Juvid on 16/7/17.
//  Copyright © 2016年 Juvid(zhutianwei). All rights reserved.
//

#import "UIView+JuLayout.h"
//#import "JuLayout.h"
@implementation UIView (JuLayout)
/*
 -(NSLayoutConstraint *)juAddConstraint:(JuLayout *)juLayout{
 self.translatesAutoresizingMaskIntoConstraints=NO;
 if (!self.superview) return nil;
 UIView *toItem=juLayout.juView2;
 UIView *addConsView=self.superview;
 CGFloat constant=juLayout.juConstant;
 if (juLayout.juAttr1==NSLayoutAttributeWidth||juLayout.juAttr1==NSLayoutAttributeHeight) {
 if (!toItem) {
 if (constant==0) {///< 等于0时约束等于父view
 toItem=self.superview;
 }else{///< 自己加约束
 addConsView=self;
 }
 }
 else if([toItem isEqual:self]){///< 长宽比例时
 addConsView=self;
 }
 }else{
 if (!toItem) {
 toItem=self.superview;
 }
 }
 NSLayoutConstraint *layoutConstraint=[NSLayoutConstraint constraintWithItem:self attribute:juLayout.juAttr1 relatedBy:juLayout.juRelation toItem:toItem attribute:juLayout.juAttr2 multiplier:juLayout.juMulti constant:constant];
 layoutConstraint.priority=juLayout.prioritys;
 layoutConstraint.juLayType=juLayout.juLayoutType;

 [self juCompareSameCons:layoutConstraint];

 if (!self.ju_Constraints) self.ju_Constraints=[NSMutableArray array];
 [self.ju_Constraints addObject:layoutConstraint];

 [addConsView addConstraint:layoutConstraint];
 return layoutConstraint;
 }*/


-(JuLayout *)newJuLayMinus:(NSLayoutAttribute)firstAtt withType:(JuLayoutType)layoutType{
    return [self newJuLayMinus:firstAtt secondAtt:firstAtt withType:layoutType];
}
-(JuLayout *)newJuLayMinus:(NSLayoutAttribute)firstAtt secondAtt:(NSLayoutAttribute)secondAtt withType:(JuLayoutType)layoutType{
    JuLayout *layout=[self newJuLay:firstAtt secondAtt:secondAtt withType:layoutType];
    layout.isMinus=YES;
    return layout;
}
-(JuLayout *)newJuLay:(NSLayoutAttribute)firstAtt withType:(JuLayoutType)layoutType{
    return [self newJuLay:firstAtt secondAtt:firstAtt withType:layoutType];
}
-(JuLayout *)newJuLay:(NSLayoutAttribute)firstAtt secondAtt:(NSLayoutAttribute)secondAtt withType:(JuLayoutType)layoutType{
    JuLayout *layout=[JuLayout new];
    layout.juAttr1=firstAtt;
    layout.juAttr2=secondAtt;
    layout.juView1=self;
    layout.juLayoutType=layoutType;
    return layout;
}
-(JuLayout *)juLead{
    return [self newJuLay:NSLayoutAttributeLeading withType:JuLayoutLead];
}
-(JuLayout *)juTrail{
    return [self newJuLayMinus:NSLayoutAttributeTrailing withType:JuLayoutTrail];
}
-(JuLayout *)juTop{
    return [self newJuLay:NSLayoutAttributeTop withType:JuLayoutTop];
}
-(JuLayout *)juBottom{
    return [self newJuLayMinus:NSLayoutAttributeBottom withType:JuLayoutBottom];
}
-(JuLayout *)juLeaSpace{
    return [self newJuLay:NSLayoutAttributeLeading secondAtt:NSLayoutAttributeTrailing withType:JuLayoutLead];
}
-(JuLayout *)juTraSpace{
    return [self newJuLayMinus:NSLayoutAttributeTrailing secondAtt:NSLayoutAttributeLeading withType:JuLayoutTrail];
}

-(JuLayout *)juTopSpace{
    return [self newJuLay:NSLayoutAttributeTop secondAtt:NSLayoutAttributeBottom withType:JuLayoutTop];
}
-(JuLayout *)juBtmSpace{
    return [self newJuLayMinus:NSLayoutAttributeBottom secondAtt:NSLayoutAttributeTop withType:JuLayoutBottom];
}
-(JuLayout *)juLastLine{
    return [self newJuLay:NSLayoutAttributeLastBaseline withType:JuLayoutBottom];
}

-(JuLayout *)juFirLine{
    return  [self newJuLay:NSLayoutAttributeFirstBaseline withType:JuLayoutTop];
}
-(JuLayout *)juCenterX{
    return [self newJuLay:NSLayoutAttributeCenterX withType:JuLayoutCenterX];
}
-(JuLayout *)juCenterY{
    return [self newJuLay:NSLayoutAttributeCenterY withType:JuLayoutCenterY];
}

-(JuLayout *)juWidth{
    return [self newJuLay:NSLayoutAttributeWidth withType:JuLayoutWidth];
}
-(JuLayout *)juHeight{
    return [self newJuLay:NSLayoutAttributeHeight withType:JuLayoutHeight];
}
-(JuLayout *)juAspectWH{
    return [self newJuLay:NSLayoutAttributeWidth secondAtt:NSLayoutAttributeHeight withType:JuLayoutAspectWH];
}


@end

File: JuLayout/UIView+JuLayout.swift
//
//  UIView+JuLayout.swift
//  JuLayout
//
//  Created by Juvid on 2016/9/16.
//  Copyright © 2016年 Juvid(zhutianwei). All rights reserved.
//

import UIKit
extension UIView{
    func newJuLayMinus(_ firstAtt:NSLayoutAttribute, layoutType:JuLayoutType) -> JuLayout {
        return self.newJuLay(firstAtt, secondAtt: firstAtt, layoutType: layoutType)
    }
    func newJuLayMinus(_ firstAtt:NSLayoutAttribute,secondAtt:NSLayoutAttribute, layoutType:JuLayoutType) -> JuLayout {
        let layout = self.newJuLay(firstAtt, secondAtt: secondAtt, layoutType: layoutType)
        layout.isMinus = true
        return layout
    }
    func newJuLay(_ firstAtt:NSLayoutAttribute, layoutType:JuLayoutType) -> JuLayout {
        return self.newJuLay(firstAtt, secondAtt: firstAtt, layoutType: layoutType)
    }

    func newJuLay(_ firstAtt:NSLayoutAttribute,secondAtt:NSLayoutAttribute, layoutType:JuLayoutType) -> JuLayout {
        let layout = JuLayout.init()
        layout.juAttr1 = firstAtt
        layout.juAttr2 = secondAtt
        layout.juView1 = self
        layout.juLayoutType=layoutType
        return layout

    }

    var juLead: JuLayout {
        return self.newJuLay(NSLayoutAttribute.leading, layoutType: JuLayoutType.lead)
    }
    var juTrail: JuLayout {
        return self.newJuLayMinus(NSLayoutAttribute.trailing, layoutType: JuLayoutType.trail)
    }
    var juTop: JuLayout {
        return self.newJuLay(NSLayoutAttribute.top, layoutType: JuLayoutType.top)
    }
    var juBottom: JuLayout {
        return self.newJuLayMinus(NSLayoutAttribute.bottom, layoutType: JuLayoutType.bottom)
    }
    var juLeaSpace: JuLayout {
        return self.newJuLay(NSLayoutAttribute.leading,secondAtt: NSLayoutAttribute.trailing, layoutType:JuLayoutType.lead)
    }
    var juTraSpace: JuLayout {
        return self.newJuLayMinus(NSLayoutAttribute.trailing,secondAtt: NSLayoutAttribute.leading, layoutType:JuLayoutType.trail)
    }
    var juTopSpace: JuLayout {
        return self.newJuLay(NSLayoutAttribute.top,secondAtt: NSLayoutAttribute.bottom, layoutType:JuLayoutType.top)
    }
    var juBtmSpace: JuLayout {
          return self.newJuLayMinus(NSLayoutAttribute.bottom,secondAtt: NSLayoutAttribute.top, layoutType:JuLayoutType.bottom)
    }
    var juLastLine: JuLayout {
        return self.newJuLay(NSLayoutAttribute.lastBaseline, layoutType: JuLayoutType.bottom)
    }
    var juFirLine: JuLayout {
        return self.newJuLay(NSLayoutAttribute.firstBaseline, layoutType: JuLayoutType.top)
    }
    var juCenterX: JuLayout {
        return self.newJuLay(NSLayoutAttribute.centerX, layoutType: JuLayoutType.centerX)
    }
    var juCenterY: JuLayout {
        return self.newJuLay(NSLayoutAttribute.centerY, layoutType: JuLayoutType.centerY)
    }
    var juWidth: JuLayout {
        return self.newJuLay(NSLayoutAttribute.width, layoutType: JuLayoutType.width)
    }
    var juHeight: JuLayout {
        return self.newJuLay(NSLayoutAttribute.height, layoutType: JuLayoutType.height)
    }
    var juAspectWH: JuLayout {
        return self.newJuLay(NSLayoutAttribute.width, secondAtt:NSLayoutAttribute.height, layoutType:JuLayoutType.aspectWH)
    }

    
    /// 设置内容优先级
    func juHorConHugPri(_ priority: UILayoutPriority) {
        self.setContentHuggingPriority(priority, for: UILayoutConstraintAxis.horizontal)
    }
    func juVerConHugPri(_ priority: UILayoutPriority) {
        self.setContentHuggingPriority(priority, for: UILayoutConstraintAxis.vertical)
    }
    func juVerConComResPri(_ priority: UILayoutPriority) {
        self.setContentCompressionResistancePriority(priority, for: UILayoutConstraintAxis.vertical)
    }
    func juHorConComResPri(_ priority: UILayoutPriority) {
        self.setContentCompressionResistancePriority(priority, for: UILayoutConstraintAxis.horizontal)
    }
//    func juLead() -> UIView {
//        return self
//    }
}
